### ISSUE TITLE
Cstring(NULL) does not equal C_NULL

### ISSUE THREAD

**Issue by maleadt** [JuliaLang/julia#13974]
It seems that a `Cstring` (according to the docs nothing more than a checked alias of `Ptr{UInt8}`) containing NULL, does not compare well to `C_NULL`:

```
julia> Ptr{UInt8}(0x0) == C_NULL
true

julia> Cstring(Ptr{UInt8}(0x0)) == C_NULL
false
```

Back-story: I was [looking into](https://groups.google.com/forum/#!topic/julia-dev/Syb7p0jg4JU) converting `Ptr{UInt8}` instances to `Cstring` in order to improve readability of some code in base. When doing so for the *return value type* of a `ccall` calling a function which returns `NULL` upon failure, the subsequent `rv == C_NULL` check now doesn't work anymore. Example of such an API: `tempnam`. I'm not sure if `Cstring` is supposed to be used in this manner though.

Tested on the latest Julia from git master.

**Comment by stevengj** [comment]
Probably we should add an `==` method for comparing `Cstring` and pointers.  Also a `Cstring(C_NULL)` constructor would be useful.

**Comment by stevengj** [comment]
`Cstring` was originally intended only for argument types, not return types, however.

**Comment by maleadt** [comment]
I assumed it was, but on the other hand it makes interactions slightly more readable (i.e. this method returns a string, rather than a random pointer).

**Comment by stevengj** [comment]
Sure, it makes sense, hence #12889.  A PR supporting pointer comparisons also makes sense.   Probably we're also missing a few other useful operations.

**Comment by maleadt** [comment]
Fixed now that #14024 is merged.
